packaging: vendor apparmor3 for improved cross-distro/platform support and easier ongoing maintenance

alexmurray opened the pull request:
Currently snap-confine relies on being compiled against libapparmor at build-time, and at runtime snapd requires the host to provide an apparmor_parser to load the generated policy. As such, snapd must generate policy using the minimum subset of features so as to avoid having the policy fail to compile on platforms which use an older apparmor_parser. By vendoring the apparmor userspace code into snapd, we can both supply a suitable libapparmor for snap-confine and ship the corresponding apparmor_parser internal to snapd. This then opens the door for future changes where we can reliably support newer policy features (such as being able to specify binding to anonymous unix sockets) and the like.

alexmurray commented:
FYI - I didn't notice that the initial commit where I checked in the apparmor-3.0.1 code missed a bunch of files due to .gitignore - so I just fixed this up and force-pushed this branch.

mvo5 commented:
Hey Alex! Thanks for looking into this. I think we need a meeting to discuss a bit how we can do this, the goal is great but I would like to discuss various approaches like using a snapcraft part and similar. Also we need to consider that some distros do not approve of vendoring like this. I will setup something :)

jhenstridge commented:
If we are vendoring the code, would it make sense to do it as a git submodule referencing the main AppArmor repository?  That would avoid copying the entire codebase into the repository, and also make it easier to move forward to new releases.

alexmurray commented:
@mvo @pedronis as discussed I have reworked this to instead only vendor apparmor into the snapd snap, and to do this via a separate part in the snapcraft.yaml. This has the downside compared to the previous approach that now snap-confine is compiled against libapparmor-dev but then we vendor in an apparmor-3.0.1 apparmor_parser into snapd. So I still have to rework the snapcraft.yaml so we can try and build the apparmor snapcraft part before the snapd-deb part and then try and set things up so that we use the vendored libapparmor part to compile the snapd-deb against.

On the upside though this means I have dropped all the previous changes to debian/rules for now so that we shouldn't introduce issues with unwanted vendoring on distros like Debian etc.

Anyway, let me know if this is more what you had in mind and I'll keep plugging away to get snapd-deb compiled against this vendored apparmor and getting any failing tests to pass etc as well.

mvo5 commented:
@alexmurray Thanks for updating this! I like the direction, I think you are right about that this needs to be tweaked so that during the snapd part build the apparmor part is used to link to snap-confine. I have not looked in-depth into the go bits yet but the direction looks sensible. I hope eventually we can always ensure to have the snapd snap so that there is no need to have the "internal" return value (because we know it's always internal).

anonymouse64 commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -127,3 +128,26 @@ parts:
+      ./autogen.sh
+      ./configure --prefix=/usr
+      make
+      # place libapparmor into staging area for use by snap-confine
+      make -C src install DESTDIR=$SNAPCRAFT_STAGE
```
did you confirm that this is sufficient to make snap-confine use the apparmor from this part? I don't remember all the details of how the deb is built, but since we aren't using a native snapcraft build system, I'm concerned that the deb build for snapd ignores whatever is in $SNAPCRAFT_STAGE (unlike if we were using one of the native snapcraft build plugins, which should pick up stuff in $SNAPCRAFT_STAGE properly).

alexmurray replied:
Yes - this is why I added 75c43fc84b5ddff32c9a55ba18577cc2c6bd4f89 which changes configure.ac to look for apparmor-3 specifically when it detects it is building via snapcraft.

anonymouse64 replied:
ack, thanks I just didn't see anything in the configure.ac changes which actually uses the $SNAPCRAFT_STAGE env var, but probably snapcraft sets some other env var that configure.ac uses to find apparmor 3 inside $SNAPCRAFT_STAGE.

anonymouse64 commented on sandbox/apparmor/apparmor.go:
```diff
@@ -316,27 +321,38 @@ func probeKernelFeatures() ([]string, error) {
+// FindAppArmorParser returns the path of the apparmor_parser binary if one is found.
+func FindAppArmorParser() (string, bool, error) {
+	// first see if we have our own internal copy
+	if path, err := snapdtool.InternalToolPath("apparmor_parser"); err == nil {
```
so if I understand correctly, if we are executing from `/usr/libexec/snapd/snapd` on i.e. fedora which disables re-exec, snapd is from the rpm package of snapd and this will only check for apparmor_parser existing in `/usr/libexec/snapd/apparmor_parser`, which will probably never exist (note this assumes that apparmor is a thing on fedora which AIUI, it isn't but just for the sake of example). And specifically this InternalToolPath will not search the snapd snap since snapd is not running from the snapd snap.

I think that's the expected and desired behavior, as on these other distros which specifically have disabled re-exec, we want to use the apparmor_parser that is on the $PATH and not from inside the snapd snap.

alexmurray replied:
Yes this is designed to prefer the snapd-snap's internal version of apparmor_parser if it is present, but otherwise fall-back to using the one from PATH so that we don't break this existing use-case.

anonymouse64 replied:
great, thanks for confirming

anonymouse64 commented on interfaces/apparmor/apparmor.go:
```diff
@@ -115,7 +116,24 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
+		// when using the internal apparmor_parser also use it's
+		// own configuration and includes etc
+		prefix := strings.TrimSuffix(parser, "apparmor_parser")
+		args = append(args, "--config-file")
+		args = append(args, filepath.Join(prefix, "/apparmor/parser.conf"))
+		args = append(args, "-b")
+		args = append(args, filepath.Join(prefix, "/apparmor.d"))
```
does this mean that the apparmor config that is used from the snapd snap is always read-only? i..e. one can't modify the parser config or add/modify an abstraction without rebuilding the snapd snap? That seems a bit unfortunate and may have implications for folks who are currently doing this to support NFS homes or other random things we don't properly support but can be supported by changing the apparmor abstractions on their system

alexmurray replied:
Yes this is perhaps unfortunate in that case, but I wanted to ensure that if we are invoking the snapd internal apparmor_parser that we then also use the associated snapd-internal apparmor config/abstractions etc. I guess the other way to do this would be to have the snapd snap copy out these files to somewhere writable when it is installed (/var/lib/snapd/apparmor/ ?) and then use this path instead here so that users can change these files if desired.

anonymouse64 replied:
Probably for now it's fine to have static config etc. from $SNAP, but this is something to think about and call out, if not eventually change. I think the proposal to have some hook/process inside snapd copy the static read-only files out of the snapd snap and into somewhere like /var/lib/snapd/apparmor is very reasonable FWIW

mardy replied:
As far as abstractions are concerned, multiple include paths can be specified, so that part should be easier to solve.

jhenstridge replied:
The apparmor_parser docs seem to indicate that `-b` and `-I` affect different types of includes (relative paths vs. absolute). It also sounds like there can only be one "base" directory.

pedronis replied:
if I understand this is a functionality regression so we do need an answer before this goes to stable? or am I confused?

alexmurray replied:
Yes this is a functionality regression (ie that now snapd will always use a static, read-only apparmor configuration which it ships itself whereas before it would use the one on the host which could then be modified etc as needed by a local admin).

But I am not sure where if anywhere we actually document this functionality - if we do publicly document this somewhere then we would want to make sure this PR can support that still in some form. Also there are advantages to shipping this static config as it means snapd should be more robust since an admin can't inadvertently mess up the global apparmor config and hence break snapd confinement etc - so I suppose we need to make a choice here whether this is something we still want to support and if so, how best to do it.

anonymouse64 replied:
Well I don't think we have any public docs on doing this, but there are various forum posts where folks have figured out that editing these config files enables them to use snapd with their specific setup, see for example https://forum.snapcraft.io/t/how-can-i-use-snap-when-i-dont-use-home-user/3352 but I don't think those comments qualify as "supported documented features".

jhenstridge replied:
As @anonymouse64 said, we've pointed people to the AppArmor tunables as ways to get snaps to work on their systems in the past. If we're changing how policies are compiled, it would be nice if those customisations continued to work. If that's not possible, I think we'd need to be able to provide a new advice to get snapd working in those environments.

Would it make sense to symlink `.../etc/apparmor.d/tunables` (or some files or directories under there) to the equivalent locations in the real `/etc`? This would require that there still be an AppArmor installation on the host system, but that's probably going to be the base for any cases where we want to install policies.

I guess the only question is whether the default tunables from old versions of AppArmor are going to be compatible with the rest of the `apparmor.d/abstractions` rules from the new AppArmor.

alexmurray replied:
I don't think we can reliably mix-and-match configurations between what we ship and what may be on the host. I think the better option is to copy out our known-good static configuration to a writable location (`/var/lib/snapd/apparmor/` ?) and then document this as where users need to customise rather than trying to include any existing host apparmor config bits. Whilst this then means they have to configure the same thing in two places, this should ensure there is less chance that some incompatibility between the snapd and the hosts apparmor config/abstractions etc causes things to break.

anonymouse64 replied:
I think it would be okay if we agreed on a timeframe when this feature could be worked on as well as who would work on implementing the feature to copy this config directory out and point apparmor_parser at that writable directory and merge this PR as-is with the understanding that the feature would not get released to stable until that feature was implemented and properly tested with i.e. spread tests and then also officially documented as something we commit to continuing to maintain.

anonymouse64 replied:
I think it's okay if in the short term we break the small quantity of users on the edge channel who may rely on the existing functionality as long as we have a documented, supported way forward for all the users currently using this workaround on the stable channel, though we no idea how many users that is

alexmurray replied:
See the addition of https://github.com/snapcore/snapd/pull/9939/commits/9b2a2866cc6024a11887011a21e435b95be84ce1 which allows the site-local tunables to be customised via a `/var/lib/snapd/apparmor/snap-tunables` file which will get included in all generated profiles IIF it exists (ie this has to be manually created by a site admin, we don't ship one by default). This should allow to alleviate any concerns with the read-only apparmor config/tunables etc inside the snapd snap as a result of this change.

anonymouse64 reviewed (commented):
not a full review, but some questions. I like the fact that we can make re-exec'ing systems move faster w/ respect to apparmor features with this, but I'm concerned about all the apparmor config being read-only and also it's a bit unclear to me that the snap build changes are sufficient to accomplish building snap-confine against the vendored apparmor in the snapd snap

anonymouse64 commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -127,3 +128,26 @@ parts:
+      cd ../../parser
+      make
+      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/lib/snapd
+      cp -a apparmor_parser $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/
```
thinking about this some more, what runtime dependencies does apparmor_parser have? is it possible that when the snapd snap is installed on trusty for example and glibc and other things are older than the build-base of snapd snap (which is currently xenial but this may change in the future) that apparmor_parser fails from the snapd snap fails to run because we don't ship/use any runtime dependencies for apparmor in the snapd snap? 

I suppose we do ship libc for the non-static things in snapd like fc-cache, so perhaps it's just a matter of adjusting how apparmor_parser is called in the internal case to make sure that the dynamic library search path only includes things from the snapd snap, but I don't know if there are other dependencies that apparmor_parser needs which we would need to also ship with the snapd snap.

anonymouse64 replied:
@alexmurray could you comment on this?

alexmurray replied:
```
$ ldd /snap/snapd/current/usr/lib/snapd/apparmor_parser 
	linux-vdso.so.1 (0x00007ffd07170000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f8fcb24a000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f8fcb0fc000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f8fcaf10000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f8fcb27c000)
```

This is similar to snap-confine etc:

```
$ ldd /snap/snapd/current/usr/lib/snapd/snap-confine
	linux-vdso.so.1 (0x00007ffd4abde000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007f38bfc3f000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f38bfc1d000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f38bfa31000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007f38bfa26000)
	libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007f38bf9fa000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f38bfc78000)
	libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007f38bf962000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f38bf959000)
```

anonymouse64 commented on sandbox/apparmor/apparmor_test.go:
```diff
@@ -41,6 +41,18 @@ type apparmorSuite struct{}
 
 var _ = Suite(&apparmorSuite{})
 
+func (*apparmorSuite) TestAppArmorFindAppArmorParser(c *C) {
```
we also need the inverse unit test too, namely that if apparmor_parser exists in the internal tool path that `internal` is `true`.

alexmurray replied:
I agree - I am just not sure how to add a test which runs under this circumstance - ie. a snapd snap installed from this PR - can you advise how I should structure this?

anonymouse64 replied:
sure, here I was just suggesting to mock some things in the unit tests, let me know if you need help using the MockCommand and SetRootDir etc functions to mock this for the unit test, I'll respond about spread tests separately

alexmurray replied:
Ah ok I'll see what I can cook up - cheers.

alexmurray replied:
Ok I added something in 6686892320

anonymouse64 reviewed (commented):
This looks good. I have another wondering about the dependencies of apparmor_parser and using it on an older system than where we build the snapd snap (which is currently build on xenial), as well as some additional unit tests.

Also, I think it would be good to either extend an existing spread test or write a new one which validates that when using the snapd snap on i.e. trusty, that apparmor_parser still is usable. We don't need full feature parity for snapd on trusty, but we will for xenial when the snapd snap migrates off of building on xenial (which I honestly don't know when that will happen, so maybe it's not really a big deal), as xenial has another 5 years of full snapd feature support (for ESM customers at least).

alexmurray commented:
@anonymouse64 thanks for the follow-up review - I agree that some more detailed tests are needed but I haven't had much luck figuring out how to add a spread test which can build and install a snapd snap from this PR and get it running locally (I don't have a google cloud credential so was just trying to run them via qemu) - can you give any guidance? Or is there some documentation I should have read that I have missed that already covers this sort of thing? This is the first time I've looked at adding new spread tests so any help you or others could give on this would be awesome (I had asked a couple times in IRC and mattermost but nobody responded there :disappointed:)

anonymouse64 commented:
@alexmurray have a look at [HACKING.md](https://github.com/snapcore/snapd/blob/master/HACKING.md) for some information on running spread tests locally with QEMU. Running spread tests against UC20 locally is a bit tricky since you need to build your own spread binary, but for all the other systems, you should be to just install spread as explained there, then create the system QEMU image you need and finally run the spread test you want to run locally with spread like so:

```
$ spread -debug -v qemu:ubuntu-18.04-64:tests/main/snapd-snap
```

As far as adding new spread tests goes, I usually just end up copying a simple spread test's `task.yaml` and modify it as needed, you can also have a look at the spread documentation [here](github.com/snapcore/spread) for a full run-down on all the fields and stuff there.

We try to make sure that HACKING.md file is up to date and usable for folks, so please have a look and let us know if anything goes wrong or doesn't work the way it is documented, we can help you out and get that document updated for others too.

anonymouse64 reviewed (commented):
A couple of suggestions for the unit test, as well as some comments on the spread test. I will file a separate PR making the necessary changes for you to the snapd-snap test to help move this along.

alexmurray commented on tests/main/security-internal-apparmor/task.yaml:
```diff
@@ -0,0 +1,39 @@
+summary: Check internal vendored apparmor in snapd snap
```
I was thinking something like that but then I wondered if it was a bit of a misuse of the existing snapd-snap test to have it do a bunch of apparmor related bits - however other than duplicating it into this test I couldn't see a way to have one test use a snapd snap built from this PR.

But the idea of being able to test a snapd snap on various releases via lxd would be awesome - so thanks for working on that - then I can drop this one from here and put those bits in that test as you suggest.

anonymouse64 replied:
The PR is here: https://github.com/snapcore/snapd/pull/10099

anonymouse64 replied:
Well it took a while to land that PR, but now we have a PR which builds the snapd snap on various ubuntu-* systems at least (there are issues building with snapcraft and LXD on other distros), so you could add some of your desired tests from here to that tests/main/snapd-snap/task.yaml too

alexmurray replied:
Yep I added something already in https://github.com/snapcore/snapd/pull/9939/commits/8c77c3242baa302b396e653f121bc60633b027f4

anonymouse64 replied:
I think this particular spread test is somewhat redundant, because even if after this PR is merged, the test will not exercise the PR, it will simply be exercising the snapd snap on stable, which may be a useful/valid test but I think in this case the test should be a nightly test instead in the nightly suite under `tests/nightly`.

alexmurray replied:
I moved this test to nightly in https://github.com/snapcore/snapd/pull/9939/commits/e3f4b30f6f4e6dfe17c627e6290a86d3a0a969ac but have left it as manual for now since until this is actually merged to master it will not pass - once this gets merged then we can remove the manual declaration.

alexmurray commented on sandbox/apparmor/apparmor_test.go:
```diff
@@ -41,6 +42,36 @@ type apparmorSuite struct{}
+	c.Assert(ioutil.WriteFile(p, nil, 0755), IsNil)
+	restore := snapdtool.MockOsReadlink(func(string) (string, error) {
+		return p, nil
+	})
```
Thanks - done in 87b8c30132

anonymouse64 commented:
closing and re-opening this so the nested tests run

anonymouse64 commented on tests/main/snapd-snap/task.yaml:
```diff
@@ -154,3 +154,18 @@ execute: |
+    unsquashfs -ll snapd_spread-test.snap | MATCH usr/lib/snapd/apparmor_parser
+
+    echo "Ensure we can install the snapd snap"
+    snap set core experimental.snapd-snap=true
```
this is not needed

alexmurray replied:
Thanks, I removed this in https://github.com/snapcore/snapd/pull/9939/commits/56a4c7a73c91b32137b41d9466a023c05f282084

anonymouse64 reviewed (commented):
lgtm, but I think the security-internal-apparmor test should move to the nightly suite instead. 

Also please unmark this PR as a RFC and update the description

anonymouse64 commented on tests/nightly/security-internal-apparmor/task.yaml:
```diff
@@ -0,0 +1,41 @@
+    snap install test-snapd-setpriority
+
+    echo "Enable installing the snapd snap, this happens automatically"
+    snap set core experimental.snapd-snap=true
+    # give the state time to create the change and then start watching
+    # it
+    retry -n 30 snap watch --last=transition-to-snapd-snap
```
this is also unnecessary, you can just do `snap install snapd --edge` these days and it "just works" everywhere

anonymouse64 commented on tests/nightly/security-internal-apparmor/task.yaml:
```diff
@@ -0,0 +1,41 @@
+summary: Check internal vendored apparmor in snapd snap
+
+manual: true
```
actually I think you can remove this, as soon as this PR is merged, the snapd snap will start being built with apparmor in it, and thus this spread test should start passing in the nightly test runs, @sergiocazzolato do you agree?

anonymouse64 replied:
ping @sergiocazzolato ^

anonymouse64 replied:
I suppose it is fine to leave this as manual and open a followup PR immediately after making it not manual. Also fwiw to respond to this comment you need to not have any pending comments as part of a "review" and then you need to find the original comment/review that I left this on and respond there for some reason, I agree it is silly and unhelpful UI, so I will just resolve this comment for the sake of simplicity

alexmurray commented:
I've updated the snapd-snap test to essentially duplicate the tests from `interfaces-many-core-provided` to try and exercise the vendored apparmor_parser etc as much as possible - and everything looks good. @mvo @pedronis what can we do to get this merged from here?

bboozzoo commented on interfaces/apparmor/apparmor.go:
```diff
@@ -117,9 +118,32 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
+		// policy features across the widest array of kernels
+		// versions
+		prefix := strings.TrimSuffix(parser, "apparmor_parser")
+		args = append(args, "--config-file")
+		args = append(args, filepath.Join(prefix, "/apparmor/parser.conf"))
+		args = append(args, "-b")
+		args = append(args, filepath.Join(prefix, "/apparmor.d"))
+		args = append(args, "--policy-features")
+		args = append(args, filepath.Join(prefix, "/apparmor.d/abi/3.0"))
```
```suggestion
		args = append(args, []string{
			"--config-file", filepath.Join(prefix, "/apparmor/parser.conf"),
			"--base", filepath.Join(prefix, "/apparmor.d"),
			"--policy-features", filepath.Join(prefix, "/apparmor.d/abi/3.0"),
		}...)
```

alexmurray replied:
That looks much nicer - thanks - https://github.com/snapcore/snapd/pull/9939/commits/35acd6d3265a3aa7489c1e8e45735038807073bb

bboozzoo commented on sandbox/apparmor/apparmor.go:
```diff
@@ -316,27 +321,38 @@ func probeKernelFeatures() ([]string, error) {
-// findAppArmorParser returns the path of the apparmor_parser binary if one is found.
-func findAppArmorParser() (string, error) {
+// FindAppArmorParser returns the path of the apparmor_parser binary if one is found.
+func FindAppArmorParser() (string, bool, error) {
```
```suggestion
func FindAppArmorParser() (path string, internal bool, err error) {
```

alexmurray replied:
Thanks - https://github.com/snapcore/snapd/pull/9939/commits/35acd6d3265a3aa7489c1e8e45735038807073bb

bboozzoo commented on sandbox/apparmor/apparmor.go:
```diff
@@ -316,27 +321,38 @@ func probeKernelFeatures() ([]string, error) {
-func findAppArmorParser() (string, error) {
+// FindAppArmorParser returns the path of the apparmor_parser binary if one is found.
+func FindAppArmorParser() (string, bool, error) {
+	// first see if we have our own internal copy
```
```suggestion
	// first see if we have our own internal copy which could come from
	// the snapd snap (likely) or be part of the snapd distro package (unlikely)
```

alexmurray replied:
https://github.com/snapcore/snapd/pull/9939/commits/35acd6d3265a3aa7489c1e8e45735038807073bb

bboozzoo commented on sandbox/apparmor/apparmor.go:
```diff
@@ -316,27 +321,38 @@ func probeKernelFeatures() ([]string, error) {
+func FindAppArmorParser() (string, bool, error) {
+	// first see if we have our own internal copy
+	if path, err := snapdtool.InternalToolPath("apparmor_parser"); err == nil {
+		if _, err := os.Stat(path); err == nil {
```
This could use `osutil.IsExecutable()` which doubles an existing check in InternalToolPath for specific scenarios, but at least we'll know if the permission bits are right in other scenarios.

alexmurray replied:
Good point - added in https://github.com/snapcore/snapd/pull/9939/commits/35acd6d3265a3aa7489c1e8e45735038807073bb

bboozzoo commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -127,3 +128,26 @@ parts:
+    override-build: |
+      cd libraries/libapparmor
+      ./autogen.sh
+      ./configure --prefix=/usr
```
probably also worth adding --disable-man-pages --disable-perl --disable-python --disable-ruby

alexmurray replied:
Thanks - done in https://github.com/snapcore/snapd/pull/9939/commits/35acd6d3265a3aa7489c1e8e45735038807073bb

alexmurray commented:
Ping @mvo5 - what else can I do to try and get this merged?

anonymouse64 commented:
@alexmurray FYI there are merge conflicts here that need to be resolved

alexmurray commented:
> @alexmurray FYI there are merge conflicts here that need to be resolved

Ah thanks @anonymouse64  - rebased on master to resolve merge conflicts.

alexmurray commented:
Since rebasing on master it looks the addition of the qualcomm-ipc-router interface is causing the snapd-snap spread test to fail when we try to connect this interface on *all* ubuntu releases when using the snapd snap - which is really strange since the whole point of this PR is to ship our own apparmor_parser and hence one which should support this apparmor feature - I'm investigating this now so hopefully this doesn't hold things up too long on my end.

alexmurray commented:
Ok so I have figured out what's going on with `qipcrtr` and the vendored apparmor - since the snapd snap is compiled on xenial, the apparmor-3 which is vendored into snapd gets compiled on xenial. When apparmor is compiled it introspects the system headers to determine what socket types are known - libc on xenial doesn't support the `qipcrtr` socket type so hence apparmor gets compiled without this support - so then no matter what platform we end up running the snapd-snap on, apparmor vendored into snapd doesn't support this socket type.

The simplest fix would be to hack the build of apparmor during the snap build to manually include this socket BUT this seems a bit fragile. Perhaps the better option would be to compile the snapd snap against the HWE kernel in xenial - then at least this will get the 4.15 kernel from bionic which should then support this... or wait until snapd gets ported to core18/core20?

alexmurray commented:
Nope, installing `linux-generic-hwe-16.04` inside the build environment is not enough - xenial would need an update for glibc to support `AF_QIPCRTR` for apparmor to be able to detect this at build time without some gross hack...

alexmurray commented:
Ok to fix the issue of unknown address families in apparmor when compiled on xenial I have vendored a copy of `af_names.h` generated on Ubuntu 21.10 so that includes the widest array of families on any system. I think this is ready for final review and merging :crossed_fingers:

pedronis commented on interfaces/apparmor/apparmor.go:
```diff
@@ -117,9 +118,31 @@ func loadProfiles(fnames []string, cacheDir string, flags aaParserFlags) error {
-	args = append(args, fnames...)
 
-	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
+	parser, internal, err := apparmor_sandbox.FindAppArmorParser()
+	if err != nil || !internal {
+		// if we couldn't find the parser with apparmor_sandbox
+		// then fall-back to trying to find one in the current PATH
+		// - same for if we are not using the internal
+		// apparmor_parser so that we can support a mocked parser
+		// during tests
+		parser = "apparmor_parser"
+	} else {
+		// when using the internal apparmor_parser also use it's
+		// own configuration and includes etc plus also ensure we
+		// use the 3.0 feature ABI to get the widest array of
+		// policy features across the widest array of kernels
+		// versions
+		prefix := strings.TrimSuffix(parser, "apparmor_parser")
+		args = append(args, []string{
+			"--config-file", filepath.Join(prefix, "/apparmor/parser.conf"),
+			"--base", filepath.Join(prefix, "/apparmor.d"),
+			"--policy-features", filepath.Join(prefix, "/apparmor.d/abi/3.0"),
+		}...)
+	}
+
+	args = append(args, fnames...)
+	output, err := exec.Command(parser, args...).CombinedOutput()
```
it feels like maybe this code belongs to a sandbox/apparmor.RunParser helper?

alexmurray replied:
I thought perhaps having the sandbox specify any required arguments would be a more flexible approach - see https://github.com/alexmurray/snapd/commit/9411f94271c6fac36f2580440431f82b1447bc0d - but if you would prefer to have the sandbox do the exec via a helper etc I'm easy either way - if we did add a `RunParser()` method, would you want that to also encapsulate the other parameters which we currently have in interfaces/apparmor like `no-expr-simplify` etc - https://github.com/snapcore/snapd/blob/062ef6a1ace0456ff21c3e89ea524dd326ac7db6/interfaces/apparmor/apparmor.go#L102 ?

alexmurray replied:
Regarding the `#include if exists` - that is a good idea but since this is in the base-template it would get included in all profiles - if we are going to take this approach, would we also want to take the opportunity to allow per-profile includes as well? ie we could append to the bottom of the snap profile `#include if exists "/var/lib/snapd/apparmor/profiles/local/snap.SNAP_NAME.APP"` ?

The other option is to do what I suggested earlier and have snapd copy out the now vendored apparmor config etc to `/var/lib/snapd/apparmor/` so that it can be modified directly (but then I wonder how we would handle the case where we want to upgrade this config in some future where we update to say apparmor-3.1 when it comes along...)

pedronis replied:
> Regarding the #include if exists - that is a good idea but since this is in the base-template it would get included in all profiles 

that's the intention, we would place the include such that it makes it easy to override things that come from the tunables

> - if we are going to take this approach, would we also want to take the opportunity to allow per-profile includes as well? ie we could append to the bottom of the snap profile #include if exists "/var/lib/snapd/apparmor/profiles/local/snap.SNAP_NAME.APP"

it seems two slightly unrelated problems. I know we ask people to play with per-snap profiles to test things, but this instead seems very upfront and maybe prone to bad habits

alexmurray replied:
The other issue with `include if exists` is that it is only supported by apparmor >= 2.13 - so this would not work on bionic or older which ship apparmor 2.12. As such I think we would only want to generate this when using the vendored apparmor but then in that case I think we should not refer to `/etc/apparmor.d/` but instead the snapd specific path of `/var/lib/snapd/apparmor/snap-tunables` or similar.

alexmurray replied:
To avoid any issues, I have added support for probing appamor_parser to see if it supports `include if exists` (https://github.com/snapcore/snapd/pull/9939/commits/02e0dc62246fdc86b108968d86739ca8b40f11cb) and only generating this rule if that is the case (https://github.com/snapcore/snapd/pull/9939/commits/9b2a2866cc6024a11887011a21e435b95be84ce1).

@pedronis + @mvo5 - let me know if there is anything else I can do to get this merged.. :crossed_fingers:

pedronis reviewed (commented):
lgtm, a code org comment that could be addressed also in a follow up possibly.

Wouldn't adding something to the common template after the tunables include like:

```
#include if exists "/etc/appamor.d/snap-tuning"
```

give us a way for people to workaround special local setups incompatibilities?

alexmurray commented on build-aux/snap/apparmor/af_names.h:
```diff
@@ -0,0 +1,223 @@
+#ifndef AF_UNSPEC
```
This is generated when building AppArmor - but it gets generated against the libc headers which are available at build time. Since snapd is core16 it gets built against xenial libc which doesn't support a bunch of different address families so we can't generate it at build time otherwise AppArmor never knows about those address families. So instead I generated it on an impish machine and vendored it in.

anonymouse64 replied:
thanks for explaining, could we at least add a comment to the top of this file that it was auto-generated on impish as part of building apparmor there, etc. ?

alexmurray replied:
Good idea - done in 80277ed89e

alexmurray commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -128,3 +129,30 @@ parts:
+  apparmor:
+    plugin: autotools
+    build-packages: [ bison, flex, gettext ]
+    source: https://launchpad.net/apparmor/3.0/3.0.3/+download/apparmor-3.0.3.tar.gz
```
I would rather be explicit about it since it is not impossible that a new AppArmor might introduce changes that could have subtle effects on policy etc.

pedronis replied:
I think I'm fine with explicit, but we need a policy about who/how we do the tracking

alexmurray replied:
Given the security team has a very close relationship to apparmor upstream then I think it makes most sense that we would do the tracking and updating here so I am happy to take this on myself.

alexmurray commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -128,3 +129,30 @@ parts:
+      # place libapparmor into staging area for use by snap-confine
+      make -C src install DESTDIR=$SNAPCRAFT_STAGE
+      cd ../../parser
+      # copy in a pregenerated list of network address families so that the
```
Unfortunately not that I can think of - see https://github.com/snapcore/snapd/pull/9939#discussion_r717125147

alexmurray commented on interfaces/apparmor/backend.go:
```diff
@@ -679,6 +679,14 @@ func addContent(securityTag string, snapInfo *snap.Info, cmdName string, opts in
+			features, _ := parserFeatures()
+			for _, f := range features {
+				if f == "include-if-exists" {
+					return "#include if exists \"/var/lib/snapd/apparmor/snap-tuning\""
```
Yep this would be a single file - I used this name as that is what @pedronis suggested earlier but I am more than happy to change it if we can come up with something more appropriate.

pedronis replied:
I'm happy for suggestions?  "global-snap-tuning" ?  "all-snap-tuning"?

anonymouse64 replied:
I like global-snap-tuning, but also as I said the best solution to the ambiguity is documentation so this is probably fine as-is too.

alexmurray commented on sandbox/apparmor/apparmor_test.go:
```diff
@@ -199,27 +237,43 @@ func (s *apparmorSuite) TestProbeAppArmorParserFeatures(c *C) {
+			expFeatures: []string{"include-if-exists", "unsafe"},
+		},
+		{
+			exitCodes:   []string{"0", "0", "0"},
```
Yep :) I'll try take a look for a future improvement PR.

alexmurray commented on tests/main/snapd-snap/task.yaml:
```diff
@@ -154,3 +164,131 @@ execute: |
+    unsquashfs -ll snapd_spread-test.snap | MATCH usr/lib/snapd/apparmor_parser
+
+    echo "Ensure we can install the snapd snap"
+    snap install --dangerous snapd_spread-test.snap
```
Added in 88a267f6ad0bc310326f916d8d515336695b2305

anonymouse64 replied:
see my other comment on what to do about this I think

alexmurray commented on tests/main/snapd-snap/task.yaml:
```diff
@@ -154,3 +164,131 @@ execute: |
+    "$TESTSTOOLS"/snaps-state install-local "$CONSUMER_SNAP"
+    tests.cleanup defer snap remove --purge "$CONSUMER_SNAP"
+
+    # If possible, prepare a session for the test user. On many systems this
```
Added in 88a267f6ad0bc310326f916d8d515336695b2305

anonymouse64 commented:
@alexmurray seems the spread test tests/main/apparmor-batch-reload is failing as well

Regarding the revert of snapd, I will try to come up with an example of how to do it, it actually seems like every ubuntu classic system, not just the dangerous variant on 16.04, is failing to revert snapd, probably because the default setup for spread is to use snapd from the branch via a hand-built deb rather than the snap, so indeed there is no prior version of snapd available.

alexmurray commented:
> @alexmurray seems the spread test tests/main/apparmor-batch-reload is failing as well

Thanks for catching that - fixed in 74c17d027cb4723ac8a6abb26263ccf81cf8c765

pedronis reviewed (approved):
overall this is fine with me,  but I didn't look again at all details. I commented on some recent questions as well

anonymouse64 commented:
@alexmurray I think you probably want to follow the example of the test code in tests/main/snapd-snap-removal in order to not fail the snapd-snap test after the snapd snap has been installed, it seems that actually almost none of the systems had the snapd snap before the test is run, so they all fail to revert since there is nothing to revert to. The test I mentioned above uses `snap-mgmt` to purge things. Let me know if you have any problems following that example

anonymouse64 commented on build-aux/snap/apparmor/af_names.h:
```diff
@@ -0,0 +1,232 @@
+// this file was generated on a Ubuntu impish install from the upstream
```
thanks for this comment

anonymouse64 reviewed (approved):
lgtm, see my other comment for how to fix the failing spread test with the revert I think

anonymouse64 commented:
Also I hope there are no objections to squash merging it, there are currently 50 commits, and I'm wondering when we squash this if there is a desired "final" commit message or should we just stick with the auto-generated one that has all the individual commit messages included? Typically what I will do is take the PR description as the commit message, but wanted to check before we do actually merge this.

alexmurray commented:
Ok tests look happier now - @mvo5 I think this is still waiting on your approval?

anonymouse64 commented:
Hmm, sadly there appears to be a problem with trusty here when installing the snapd snap: https://pastebin.ubuntu.com/p/pDNY5V3MNj/

Specifically something about 
```
snapd[3467]: cannot run daemon: state startup errors: [cannot obtain snap-seccomp version information: /snap/snapd/x1/usr/lib/snapd/snap-seccomp: symbol lookup error: /snap/snapd/x1/usr/lib/snapd/snap-seccomp: undefined symbol: seccomp_version]
```

unclear to me if that is related to this PR or if maybe trusty just doesn't support using the snapd snap properly? I really hope it's not the latter...

alexmurray commented:
Ok so I figured out some more bits here - the snapd debian/rules has a hack to statically link snap-seccomp against libseccomp - which is why the snapd snap build by github actions has snap-seccomp statically linked. debian/rules is a symlink to packaging/ubuntu-16.04/rules.

But the packaging/ubuntu-14.04/rules does not have this same hack - and these get used when the spread host is 14.04 - so probably the simplest way to try and workaround this is to have the snapd-snap spread test always install the packaging/ubuntu-16.04 as the debian directory before trying to build the snapd snap via LXD. Will give that a go.

alexmurray commented:
@anonymouse64 no, from what I can see, prepare-restore builds a deb first on the host - so on trusty it goes and changes the debian/ symlink to point to packaging/ubuntu-14.04 - *then* the snapd-snap test runs - so when it then runs LXD to build the snap, LXD will use the 14.04 packaging as that is what the debian/ symlink is already set to - so we need to go and undo this before building the snap.

anonymouse64 commented:
@alexmurray ah so really the bug is in our prepare-restore script, specifically these bits :

https://github.com/snapcore/snapd/blob/7531f82780a11425a7734778022aa48adfdb5af7/tests/lib/prepare-restore.sh#L348-L349

that needs to be undone by the prepare-restore.sh script before executing any test IMHO

anonymouse64 commented:
@alexmurray probably the exact same code you have now to restore the debian symlink in snapd-snap/task.yaml should really be done around line 564 of prepare-restore.sh to prevent any test from running into this issue

bboozzoo commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -128,3 +129,30 @@ parts:
+      cd libraries/libapparmor
+      ./autogen.sh
+      ./configure --prefix=/usr --disable-man-pages --disable-perl --disable-python --disable-ruby
+      make
```
Sorry for not noticing this earlier, this could be `make -j`.

bboozzoo commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -128,3 +129,30 @@ parts:
+      # parser gets built to support as many as possible even if glibc in
+      # the current build environment does not support them
+      cp $SNAPCRAFT_PROJECT_DIR/build-aux/snap/apparmor/af_names.h .
+      make
```
same here

bboozzoo commented on build-aux/snap/snapcraft.yaml:
```diff
@@ -128,3 +129,30 @@ parts:
+      mkdir -p $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor
+      cp -a parser.conf $SNAPCRAFT_PART_INSTALL/usr/lib/snapd/apparmor/
+      cd ../profiles
+      make
```
and here

bboozzoo commented on cmd/configure.ac:
```diff
@@ -86,8 +86,13 @@ AS_IF([test "x$with_unit_tests" = "xyes"], [
 AS_IF([test "x$enable_apparmor" = "xyes"], [
-    PKG_CHECK_MODULES([APPARMOR], [libapparmor], [
-      AC_DEFINE([HAVE_APPARMOR], [1], [Build with apparmor support])])
+    # Expect AppArmor3 when building as a snap
```
```suggestion
    # Expect AppArmor3 when building as a snap under snapcraft
```

bboozzoo commented on interfaces/apparmor/backend.go:
```diff
@@ -691,6 +691,14 @@ func addContent(securityTag string, snapInfo *snap.Info, cmdName string, opts in
 	}
 	policy = templatePattern.ReplaceAllStringFunc(policy, func(placeholder string) string {
 		switch placeholder {
+		case "###INCLUDEIFEXISTSSNAPTUNING###":
+			features, _ := parserFeatures()
+			for _, f := range features {
+				if f == "include-if-exists" {
+					return "#include if exists \"/var/lib/snapd/apparmor/snap-tuning\""
+				}
+			}
+			return ""
```
```suggestion
		case "###INCLUDEIFEXISTSSNAPTUNING###":
			features, _ := parserFeatures()
			if strutil.ListContains(features, "include-if-exists") {
				return `#include if exists "/var/lib/snapd/apparmor/snap-tuning"`
			}
			return ""
```

alexmurray replied:
Thanks, done in 823b25f3b5

bboozzoo commented on interfaces/apparmor/backend_test.go:
```diff
@@ -990,6 +990,44 @@ func (s *backendSuite) TestCombineSnippetsChangeProfile(c *C) {
+		expected: "",
+	}, {
+		features: []string{"include-if-exists"},
+		expected: "#include if exists \"/var/lib/snapd/apparmor/snap-tuning\"",
```
```suggestion
		expected: `#include if exists "/var/lib/snapd/apparmor/snap-tuning"`,
```

bboozzoo reviewed (commented):
Couple more comments

alexmurray commented:
Sure 2.54 sounds fine - squash merging is fine too - there are so many small commits here I don't think they make sense to keep in the final merge - as for a commit message:

```
many: Vendor apparmor-3.0.3 into the snapd snap

Vendor apparmor-3.0.3 into the snapd snap to provide a common,
supported and modern apparmor version for snapd confinement.

AppArmor 3 provides the network v8 ABI which include AF_UNIX and the
various other AppArmor features which snapd utilises for snap confinement.

This takes only the apparmor_parser binary and the associated parser.conf
configuration file for it, plus the profile abstractions, tunables and abi
definitions as this is the bare-minimum needed to support compiling
AppArmor profiles that are generated by snapd.

Also extend the snapd-snap test to perform various tests on the resulting
snap to validate AppArmor confinement when using this change.

Finally, since the snapd snap will now use our own immutable apparmor
configuration, to allow local administrators to still easily tune global snap
AppArmor policy, add support for optional site-local tunables to each
snap AppArmor profile via the include if exists support in AppArmor 3.
```

xnox commented:
Now that https://github.com/snapcore/snapd/pull/10957 is merged, I guess master needs to be remerged into here again.

alexmurray commented:
@mvo5 any chance this can get merged?

mvo5 commented:
@alexmurray I tried to merge this but after looking closely at the spread failures I see:
```
2021-11-18T12:11:37.0087983Z ##[error]2021-11-18 12:11:37 Error executing google:ubuntu-18.04-64:tests/main/lxd:snapd_cgroup_both (nov181201-442715) : 
2021-11-18T12:11:37.0118562Z -----
...
2021-11-18T12:11:37.1636118Z + MATCH from-the-inside
2021-11-18T12:11:37.1637787Z + lxd.lxc exec my-ubuntu -- su -l ubuntu -c '/snap/bin/test-snapd-sh.sh -c '\''echo from-the-inside'\'''
2021-11-18T12:11:37.1639373Z cannot change profile for the next exec call: No such file or directory
2021-11-18T12:11:37.1640662Z grep error: pattern not found, got:
2021-11-18T12:11:37.1641681Z -----
2021-11-18T12:11:37.1642229Z .
2021-11-18T12:11:38.2920458Z 2021-11-18 12:11:38 Debug output for google:ubuntu-18.04-64:tests/main/lxd:snapd_cgroup_both (nov181201-442715) : 
```
which indicates that something broke with this PR when running snaps inside an lxd container. I double checked and this test passes when running against "upstream/master" but it does fail like with "vendor-appamor3" (I merged upstream/master into it just to be sure before running the test):
```
$ $GOPATH/bin/spread -v qemu:ubuntu-18.04-64:tests/main/lxd
...
+ MATCH from-the-inside
+ lxd.lxc exec my-ubuntu -- su -l ubuntu -c '/snap/bin/test-snapd-sh.sh -c '\''echo from-the-inside'\'''
cannot change profile for the next exec call: No such file or directory
grep error: pattern not found, got:
```
It's very easy to reproduce. What is confusing is that this seems to only happen on 18.04, not on 20.04 or 16.04.

alexmurray commented:
Oh no, thanks for noticing that @mvo5 - after debugging it looks like `apparmor_parser` on 18.04 exits with 0 even when it fails to parse a profile (such as when we are probing for features). As such, the apparmor sandbox assumes all features are supported and so generates profiles containing the `#include if exists` - when we then go to load the policy it fails (silently) and hence when snap-confine goes to confine the process we see:

`cannot change profile for the next exec call: No such file or directory`

As the profile never got loaded... so I think the best way around this is to check both the exit status from apparmor_parser *as well* as looking for `parser error` in the output when probing for parser features. I'm just checking if this resolves the issue and if so will push a further commit containing that change.

jdstrand commented on interfaces/apparmor/backend.go:
```diff
@@ -704,16 +704,20 @@ func addContent(securityTag string, snapInfo *snap.Info, cmdName string, opts in
+		case "###INCLUDEIFEXISTSSNAPTUNING###":
+			features, _ := parserFeatures()
+			if strutil.ListContains(features, "include-if-exists") {
+				return `#include if exists "/var/lib/snapd/apparmor/snap-tuning"`
```
I'm super-excited that apparmor3 is now being built into the snapd snap like this. Nice! This `snap-tuning` seems like an unrelated change. How is this meant to be used? Is it documented somewhere?

anonymouse64 replied:
It's meant to be used to drop in additional apparmor rules like we recommended folks who have non-standard $HOME setups to do, since the recommendation in i.e. https://forum.snapcraft.io/t/how-can-i-use-snap-when-i-dont-use-home-user/3352 doesn't work anymore with apparmor bundled inside the snapd snap as the options we pass to apparmor_parser specifically only include dirs etc from the snapd snap and not from the host. So now the recommendation will be to put such tunables inside this specific file/directory instead.

I think we do intend to document this, but there were some unresolved questions about the support level of that file...
